Feat!: Add support for concurrent table diff of multiple models

themisvaltinos opened the pull request:
This update extends using the table_diff command with `--select-model` to diff a subset or selection of SQLMesh models, fixes: #4198 

https://github.com/user-attachments/assets/7a2c817d-8743-4ddb-abda-cb9db04fbe26


For example with the downstream + to diff `items` and all models downstream of it:
`sqlmesh table_diff prod:dev sushi.items+`  

With the * wildcard to diff all models of a schema:
`sqlmesh table_diff prod:dev 'sushi.*'`  

Similarly with all the selectors: https://sqlmesh.readthedocs.io/en/latest/guides/model_selection/

- If the --show-sample flag is included, the output also includes sample rows. 
- If the engine supports it, it runs all the table diff concurrently.

izeigerman commented on sqlmesh/core/console.py:
```diff
@@ -197,7 +197,7 @@ def show_intervals(self, snapshot_intervals: t.Dict[Snapshot, SnapshotIntervals]
 
 class DifferenceConsole(abc.ABC):
-    """Console for displaying environment differences"""
+    """Console for displaying differences"""
```
@benfdking is this how you intended for this interface to be used?

themisvaltinos replied:
talked to Ben and decided to separate these methods into a different class

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1648,15 +1649,141 @@ def table_diff(
+            )
+
+        table_diffs: t.List[TableDiff] = []
+        with ThreadPoolExecutor(
```
Please use https://github.com/TobikoData/sqlmesh/blob/main/sqlmesh/utils/concurrency.py#L241 instead

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1648,15 +1649,141 @@ def table_diff(
+
+        def run_diff(snapshot_name: str) -> TableDiff:
+            return self.table_diff(
+                source=source,
```
So are we going to fetch both environments N more times where N is the number of snapshots that are different?

themisvaltinos replied:
you're right didn't realise that, I should refactor this so it doesn't call table_diff or break it up into methods to guarantee nothing is called twice

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1599,156 @@ def table_diff(
-            model = self.get_model(model_or_snapshot, raise_if_missing=True)
-            adapter = self._get_engine_adapter(model.gateway)
+        # Diffs multiple or a single model across two environments
+        if model_or_snapshot or select_model:
```
I don't think we need model_or_snapshot any more. We only need `select_model`. In CLI if a single model name was provided we can just convert it into `select_models["<model_name>"]`

themisvaltinos replied:
revised the logic to only have select models there to handle them the same way

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1599,156 @@ def table_diff(
+
+        target = next(
+            snapshot for snapshot in target_env.snapshots if snapshot.name == model.fqn
+        ).qualified_view_name.for_environment(target_env.naming_info, adapter.dialect)
```
Now this approach may not actually work because the selector that was used to apply a plan in dev can be different from the selector used for table diffing. So views might not actually exist in the environment. I think it's fine for now but keep in mind that we might need to fix this in future.

themisvaltinos replied:
yes moved this logic so that we handle this case and we don't diff these models and inform the user as with the other details at the start of the diff

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1599,156 @@ def table_diff(
+                models_to_diff = self._new_selector().expand_model_selections(select_model)
+                target_snapshots = {
+                    s.name: s
+                    for s in self.state_reader.get_snapshots(target_env.snapshots).values()
```
We don't need the full snapshots. All we need are 2 environment records. `SnaphotTableInfo` that are in the environment record are more than enough to tell us whether snapshots are different or not. I don't think we need the full blown context diff for this.

themisvaltinos replied:
I revised this to instead compare the fingerprints so that we don't have to do the context diff

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1599,156 @@ def table_diff(
+            )
+            if select_model:
+                models_to_diff = self._new_selector().expand_model_selections(select_model)
+                target_snapshots = {
```
For models that were excluded this way we need some sort of message indicating that the same model version is used in both environments and there's no point in diffing. Otherwise, it will be confusing to the user why the models they selected haven't been compared.

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1599,156 @@ def table_diff(
+                )
+                modified_snapshots = {
+                    current_snapshot.snapshot_id.name
+                    for _, (current_snapshot, _) in context_diff.modified_snapshots.items()
```
For models that were excluded this way we need some sort of message indicating that the same model version is used in both environments and there's no point in diffing. Otherwise, it will be confusing to the user why the models they selected haven't been compared.

themisvaltinos replied:
yes good point, added an initial detail information of these models that aren't going to be diffed and which are

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1599,156 @@ def table_diff(
+                )
+            ]
+
+        if show:
```
I think this should be handled in the CLI code and the `show` argument removed from this method.

themisvaltinos commented:
Thanks for the review @izeigerman addressed comments, included messages indicating which models will be diffed and which will not and added a progress bar as well as information about which models are currently being processed for table differences, before providing the full table diffs at the end. I added a video in the description of this pr of how this look

sungchun12 commented on sqlmesh/cli/main.py:
```diff
@@ -891,18 +891,25 @@ def create_external_models(obj: Context, **kwargs: t.Any) -> None:
     help="Schema used for temporary tables. It can be `CATALOG.SCHEMA` or `SCHEMA`. Default: `sqlmesh_temp`",
 )
+@click.option(
+    "--select-model",
```
```suggestion
    "--select-model",
    "-s",
```
makes the flag look less intimidating to reach for

benfdking replied:
nit: Can we just shorten it to 

```
--model
-m
```

`m` Feels more intuitive than an `s`?

sungchun12 replied:
Then we'll have to change it for all other similar flags. I disagree for this PR.

themisvaltinos replied:
I agree with Sung since we have the selector in the other commands as well so we can't change `--select-model`. though for the shorten version I opted for `-m` since the `-s` flag is already used for `--skip-columns`

sungchun12 commented:
Can you show specific file path to models that don't have grain specified instead of this generic error?
![image](https://github.com/user-attachments/assets/63f9b5f0-f928-4bf9-a362-827889b1d2d2)


and can you prevent data diffs from running at all until all grains in models to diff are configured?

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1570,15 +1571,15 @@ def table_diff(
         skip_grain_check: bool = False,
         temp_schema: t.Optional[str] = None,
-    ) -> TableDiff:
+    ) -> t.Union[TableDiff, t.List[TableDiff]]:
```
Should we just always return list?

themisvaltinos replied:
yes, revised it to always return a list

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1598,178 @@ def table_diff(
+            models_no_diff: t.List[str] = []
+
+            for model_or_snapshot in selected_models:
+                model = self.get_model(model_or_snapshot, raise_if_missing=True)
```
`get_model` performs quite a few extra steps that you don't really need here since the selector returns FQNs. I'm pretty sure you can just do:
```
for model_fqn in selected_models:
    model = self._models[model_fqn]
    ```

themisvaltinos replied:
yes that's true we don't need to use get_model anymore

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1598,178 @@ def table_diff(
+                model = self.get_model(model_or_snapshot, raise_if_missing=True)
+                adapter = self._get_engine_adapter(model.gateway)
+                source_snapshot = next(
+                    (snapshot for snapshot in source_env.snapshots if snapshot.name == model.fqn),
```
Is O(N^2)? Should we build a mapping beforehand?

themisvaltinos replied:
good point yes in the worst case it is o(n^2) so revised to build the snapshot maps first

themisvaltinos commented:
> Can you show specific file path to models that don't have grain specified instead of this generic error? 
> and can you prevent data diffs from running at all until all grains in models to diff are configured?

yes really good points, changed to provide the file path and model names that don't have grain and also to prevent the diffs from running in that case. so it currently looks like this:

<img width="1015" alt="Screenshot 2025-05-02 at 12 02 26" src="https://github.com/user-attachments/assets/f1ec3367-6810-4730-843a-5799b2b356a9" />

izeigerman commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1598,200 @@ def table_diff(
+                            target_env.naming_info, adapter.dialect
+                        )
+
+                        model_on = []
```
This block looks like it should be done in a separate function. Maybe a method in the model definition (or wherever we define `all_references`).

sungchun12 commented:
Before you merge, I want to review the UX and hammer away it likely usage patterns. Great update so far! After, I'll approve.

sungchun12 commented on sqlmesh/core/context.py:
```diff
@@ -1597,53 +1598,184 @@ def table_diff(
+                    )
+                    raise SQLMeshError(
+                        f"SQLMesh doesn't know how to join the tables for the following models:\n{model_names}\n"
+                        "\nPlease specify the `grains` in each model definition."
```
```suggestion
                        "\nPlease specify the `grain` in each model definition."
```
exact config name

sungchun12 commented on sqlmesh/cli/main.py:
```diff
@@ -892,18 +892,26 @@ def create_external_models(obj: Context, **kwargs: t.Any) -> None:
+    "-m",
+    type=str,
+    multiple=True,
+    help="Select specific model that should be diffed.",
```
```suggestion
    help="Specify one or more models to data diff. Use wildcards to diff multiple models. Ex: '*' (all models with applied plan diffs), 'demo.model+' (this and downstream models)",
```

sungchun12 reviewed (changes requested):
Found a bug. It currently diffs models that are indirect non-breaking. These should be skipped.

```
Directly Modified: demo__dev.incremental_model (Non-breaking)
└── Indirectly Modified Children:
    ├── demo__dev.full_model (Indirect Non-breaking)
    └── demo__dev.full_model_example (Indirect Non-breaking)


Models to compare:
├── demo.customers
├── demo.full_model
├── demo.full_model_example
├── demo.incremental_model
└── demo.stg_customers
```

Noticed a UX bug when `table_diff` errors out. It makes the cursor invisible in my terminal. I had to run this to see it again.

`echo -e '\033[?25h'`

bad selector bug. This should NOT run any diffs in this scenario.

```
(.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff prod:dev -m 'git:sungvscode'
fatal: ambiguous argument 'sungvscode...': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'

Models to compare:
├── demo.stg_customers
├── demo.incremental_model
└── demo.customers


Aborted!
Calculating model differences ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 0.0% • pending • 0:00:09
    demo.stg_customers ...                                                                     
demo.incremental_model ...                                                                     
        demo.customers ... 
```

How exactly does this selector flag work? Looks like only models with direct modifications and not the downstream impacts.

```
(.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff prod:dev -m 'git:sung/vscode'

Models to compare:
├── demo.stg_customers
├── demo.customers
└── demo.incremental_model
```

When I run a selector that fails, it shows nothing. Can you put a message that says something like, "No models match this model selector"

```
(.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff prod:dev -m 'increm*' 
```
Also I found another bug, but this may be unrelated to this PR. 

```
Calculating model differences ━━━━━━━━╺━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 20.0% • 1/5 • 0:00:11
        demo.full_model .                                                                   
demo.full_model_example .                                                                   
 demo.incremental_model .                                                                   
     demo.stg_customers .                                                                   Error: 10 validation errors for RowDiff
stats.s_count
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.t_count
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.join_count
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.null_grain_count
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.full_match_count
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.item_id_matches
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.num_orders_matches
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.updated_at_matches
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.s_only_count
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
stats.t_only_count
  Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
    For further information visit https://errors.pydantic.dev/2.11/v/float_type
```

themisvaltinos commented:
Thanks for the review @sungchun12 if you want to have another look, addressed comments and updated the docs

> Can you add to the table diff guide? Something like, "Diffing multiple models at the same time"

yes added in the table diff guide details as well as example use cases of the selectors

> Found a bug. It currently diffs models that are indirect non-breaking. These should be skipped.

addressed it by comparing the data hash instead so that indirect non breaking models are skipped

> Noticed a UX bug when `table_diff` errors out. It makes the cursor invisible in my terminal. I had to run this to see it again.

good catch! added a try catch handle that in this cases properly closes the Live rich console so that the terminal settings are restored such as the cursor

> bad selector bug. This should NOT run any diffs in this scenario.
> 
> ```
> (.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff prod:dev -m 'git:sungvscode'
> fatal: ambiguous argument 'sungvscode...': unknown revision or path not in the working tree.
> Use '--' to separate paths from revisions, like this:
> 'git <command> [<revision>...] -- [<file>...]'
> 
> Models to compare:

we previously didn't handle the error that was caused by running the git cli command. revised so that we capture the error and display it in a more clean way to the user and raise an appropriate error and not diff

> How exactly does this selector flag work? Looks like only models with direct modifications and not the downstream impacts.
> ```
> (.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff prod:dev -m 'git:sung/vscode'
> 
> Models to compare:
> ├── demo.stg_customers
> ├── demo.customers
> └── demo.incremental_model
> ```

yes in the case of the git selector to get downstream of upstream dependencies of these models it should be used with the `+` selector e.g. `+git:sung/vscode+` . I see we didn't have details in the docs for these so added in the docs more details about how to combine them.

> When I run a selector that fails, it shows nothing. Can you put a message that says something like, "No models match this model selector"
> 
> ```
> (.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff prod:dev -m 'increm*' 
> ```

yes added a relevant message that also displays the selectors used

> Also I found another bug, but this may be unrelated to this PR.
> ```
> Input should be a valid number [type=float_type, input_value=None, input_type=NoneType]
> ```

yes I believe this is related to this issue: https://github.com/TobikoData/sqlmesh/pull/4310 that Erin fixed a couple of hours ago

sungchun12 commented on docs/guides/tablediff.md:
```diff
@@ -122,6 +122,77 @@ Under the hood, SQLMesh stores temporary data in the database to perform the com
+
+When diffing multiple models, SQLMesh will:
+
+1. Show which models exist only in the source environment
```
Is source prod and target is dev? 

If yes, then this message should say this for a new dev only model and is a bug.

`Models only in target environment`

```
Models only in source environment:
└── "sqlmesh-public-demo"."demo"."incremental_model_new"
```

themisvaltinos replied:
yes, this was indeed incorrect. I’ve updated it to exclude those model lists altogether, since the selector returns models from the latest environment where a plan has been applied so It can only retrieve one of the two model lists so it wouldn't be clear or coherent to the user. instead, kept to show only models that exist in both environments to keep the output more streamlined and consistent

sungchun12 commented on docs/guides/tablediff.md:
```diff
@@ -122,6 +122,77 @@ Under the hood, SQLMesh stores temporary data in the database to perform the com
+4. Compare the models that have differences and display them
+
+The `--select-model` option supports a powerful selection syntax that lets you choose models using patterns, tags, dependencies and git status. For complete details, see the [model selection guide](./model_selection.md).
+
```
> Note: models will only be data diffed if there's a breaking change that impacts them.

sungchun12 commented on docs/guides/tablediff.md:
```diff
@@ -122,6 +122,77 @@ Under the hood, SQLMesh stores temporary data in the database to perform the com
+
+When multiple selectors are provided, they are combined with OR logic, meaning a model matching any of the selectors will be included.
+
+All the standard table diff options like `--show-sample` work with multiple model diffing. The comparisons are executed concurrently for better performance when dealing with a large of project or many models.
```
```suggestion
All the standard table diff options like `--show-sample` work with multiple model diffing. The comparisons are executed concurrently for better performance when dealing with a large project or many models.
```

sungchun12 commented on docs/guides/tablediff.md:
```diff
@@ -122,6 +122,77 @@ Under the hood, SQLMesh stores temporary data in the database to perform the com
+The `--select-model` option supports a powerful selection syntax that lets you choose models using patterns, tags, dependencies and git status. For complete details, see the [model selection guide](./model_selection.md).
+
+Here are some common patterns you can use:
+
```
```suggestion
> Note: Surround your selection pattern in single or double quotes. Ex: '*', "sqlmesh_example.*"
```

sungchun12 reviewed (changes requested):
Couple suggestions on docs and one more question that may be a bug. If not a bug, I'll approve!

sungchun12 commented on docs/guides/tablediff.md:
```diff
@@ -122,6 +122,77 @@ Under the hood, SQLMesh stores temporary data in the database to perform the com
+3. Show which models exist in both environments but have no differences
+4. Compare the models that have differences and display them
+
+The `--select-model` option supports a powerful selection syntax that lets you choose models using patterns, tags, dependencies and git status. For complete details, see the [model selection guide](./model_selection.md).
```
```suggestion
The `--select-model` option supports a powerful selection syntax that lets you choose models using patterns, tags, dependencies and git status. For complete details, see the [model selection guide](./model_selection.md).

> Note: models will only be data diffed if there's a breaking change that impacts them.
```

sungchun12 commented on docs/guides/tablediff.md:
```diff
@@ -122,6 +122,77 @@ Under the hood, SQLMesh stores temporary data in the database to perform the com
+
+All the standard table diff options like `--show-sample` work with multiple model diffing. The comparisons are executed concurrently for better performance when dealing with a large of project or many models.
+
+> Note: All models being compared must have their `grain` defined, as this is used to perform the join between the tables in the two environments.
```
```suggestion
> Note: All models being compared must have their `grain` defined that is unique and not null, as this is used to perform the join between the tables in the two environments.
```

sungchun12 reviewed (approved):
Minor wording suggestions. Overall, tested the UX and feels great to use. Can't wait for this to be a canonical part of the demo soon!

georgesittas commented on sqlmesh/core/context.py:
```diff
@@ -1587,7 +1588,7 @@ def table_diff(
                 If omitted, the table's grain will be used.
             skip_columns: The columns to skip when computing the table diff.
-            model_or_snapshot: The model or snapshot to use when environments are passed in.
+            select_models: The modelσ or snapshotσ to use when environments are passed in.
```
Stray Greek sigma lol

themisvaltinos replied:
thanks for catching this!

georgesittas reviewed (commented):
This is breaking because `table_diff`'s API changed.

sungchun12 commented:
This may be too verbose and annoy users if they have hundreds of models with no differences. People will 99% care about only seeing models to compare. Happy to wait for another PR or remove this information directly in this one. 

```
(.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff dev:dev_sung -m '*'

Models without changes:
├── "sqlmesh-public-demo"."customer2"."orders"
├── "sqlmesh-public-demo"."demo"."external_model"
├── "sqlmesh-public-demo"."demo"."full_model_example"
├── "bigquery-public-data"."ga4_obfuscated_sample_ecommerce"."events_20210131"
├── "sqlmesh-public-demo"."demo"."incremental_model_new"
├── "sqlmesh-public-demo"."demo"."stg_payments"
├── "sqlmesh-public-demo"."demo"."stg_customers"
├── "sqlmesh-public-demo"."customer1"."custom"
├── "sqlmesh-public-demo"."demo"."incremental_model"
├── "sqlmesh-public-demo"."demo"."seed_ecommerce"
├── "sqlmesh-public-demo"."customer1"."orders"
├── "sqlmesh-public-demo"."demo"."seed_model"
├── "sqlmesh-public-demo"."demo"."incremental_by_unique_key_example"
├── "sqlmesh-public-demo"."demo"."stg_orders"
├── "sqlmesh-public-demo"."demo"."customers"
├── "sqlmesh-public-demo"."demo"."orders"
├── "sqlmesh-public-demo"."demo"."full_model"
├── "sqlmesh-public-demo"."demo"."incremental_forward_only_model"
├── "sqlmesh-public-demo"."customer5"."orders"
├── "sqlmesh-public-demo"."customer4"."orders"
├── "sqlmesh-public-demo"."customer3"."orders"
├── "sqlmesh-public-demo"."demo"."seed_raw_customers"
├── "sqlmesh-public-demo"."demo"."seed_raw_orders"
└── "sqlmesh-public-demo"."demo"."seed_raw_payments
```

themisvaltinos commented on docs/guides/tablediff.md:
```diff
@@ -122,6 +122,85 @@ Under the hood, SQLMesh stores temporary data in the database to perform the com
+
+> Note: Models will only be data diffed if there's a breaking change that impacts them.
+
+Here are some common patterns you can use:
```
good point I cleaned it up now as it makes sense to have them in the model selector guide. I simply kept here a really concise section of examples for table diff and simply linked for the detailed one in the model selector

themisvaltinos commented:
thanks for the review @georgesittas addressed the latest round of comments, moved the detailed selector examples in the model_selection guide, kept only a concise section in table dif and removed the verbose example "*" as @sungchun12 suggested

sungchun12 commented:
Thanks for removing the models without changes section! 

Can you make sure to include that same warning message when no models match criteria?

`No models matched the selection criteria: '1j3hia'`

Looks like for git selectors, `Models without changes:` is still showing up when it shouldn't.

The bug examples below.

```
(.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff dev:dev_sung -m '*'
(.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff dev:dev_sung -m 'git:sung/vscode'

Models without changes:
├── "sqlmesh-public-demo"."demo"."stg_customers"
├── "sqlmesh-public-demo"."demo"."customers"
├── "sqlmesh-public-demo"."demo"."incremental_model"
└── "sqlmesh-public-demo"."demo"."incremental_model_new"

(.venv) ➜  sqlmesh-demos git:(sung/vscode) ✗ sqlmesh table_diff prod:dev_sung -m '1j3hia'
No models matched the selection criteria: '1j3hia'

Models without changes:
```
